### ISSUE TITLE
`.flush()` doesn't work when pretty-print is enabled

### ISSUE THREAD

**Issue by roim** [pinojs/pino#1015]
TSIA

https://github.com/roim/pino-pretty.async-flush

**Comment by roim** [comment]
Funny enough, when you try to use final you get:

```
pino.final with prettyPrint does not support flushing
```

The limitation isn't just for final, yet regular loggers don't have a warning.

**Comment by mcollina** [comment]
Unfortunately so. We do not think it is needed either as prettyPrint is just something for development - so we did not implement it. If you would like to work on it, I'll be happy to review.

Note that https://github.com/pinojs/pino/pull/1003 will make all sort of things possible.

**Comment by roim** [comment]
I agree it's not needed (can't think of a scenario where you'd *need* pretty print AND async). This is more of an usability issue--eg it took me maybe 30 min to realize what was wrong in my pino integration.

Do you guys have a stance on how to mitigate these usability pain points? I'd be happy to either log a warning on `.flush` calls with async mode enabled (assume the perf hit there is negligible since `.flush` is already intensive), or to just add some doc hints.

**Comment by mcollina** [comment]
I'm ok with both a doc approach and a noisy one. @jsumners?

**Comment by jsumners** [comment]
I am okay with both.

Also, I do not know what "TSIA" means.

**Comment by micalevisk** [comment]
> I do not know what "TSIA" means.

_Title Says It All_ :smile:

**Comment by meehawk** [comment]
@mcollina @jsumners I'd love to work on this. This is going to be my first contribution to `pinojs`. Any pointers on what would be a good start?

**Comment by mcollina** [comment]
Update the doc of `pino.final()` telling that prettyPrint would not work. Also add it to the docs of `prettyPrint`.

**Comment by davorpa** [comment]
Hey @mcollina @roim 👋

I'm trying to do my first contribution with this. 🤗
After reading how final and asynchronous logging works, I think showcase could be already covered.

> 📑 `pino.final()` notes:
https://github.com/pinojs/pino/blob/01a8633d7ff1b59e9fa47cc47e34421053fea10e/docs/api.md#L1016-L1027

> 📑 `Asynchronous logging` notes:
https://github.com/pinojs/pino/blob/01a8633d7ff1b59e9fa47cc47e34421053fea10e/docs/asynchronous.md#L74-L108

> 📑 `pino({options.prettyPrint})` notes:
https://github.com/pinojs/pino/blob/01a8633d7ff1b59e9fa47cc47e34421053fea10e/docs/api.md#L345-L363
📑 `options.suppressFlushSyncWarning` notes:
https://github.com/pinojs/pino/blob/01a8633d7ff1b59e9fa47cc47e34421053fea10e/docs/pretty.md#L85-L100
↘ `suppressFlushSyncWarning = true`:
https://github.com/pinojs/pino/blob/01a8633d7ff1b59e9fa47cc47e34421053fea10e/lib/tools.js#L234-L247

Let me know if I'm wrong 😉.

Last note about `suppressFlushSyncWarning` seems to be the noisy approach 🤔. It's warned only on first flush. Then, how to proceed? 💡Maybe mention in `final()` and `pretty printers` to know how to disable it?

BR

**Comment by mcollina** [comment]
Doc would be good, yes! Would you like to send a PR?

**Comment by davorpa** [comment]
I'll try @mcollina.

Let me see how I can write according to the rest of the documentation

**Comment by Mange** [comment]
Some input on why I'd like a flush mode. I write the log messages to `stdout`/`stderr` using pretty, and then I want to also output other things to the same stream.

As it is right now, I cannot wait for the log message to appear before writing the other data. My own data is always mixed up with the last log line, which appears after a visible delay.

I'm currently looking at using workarounds like this:

```ts
// Trying to flush Pino does not work…
await new Promise((resolve) => logger.flush(() => resolve(null)));
// And trying to manually wait for Node to think it all has been written doesn't help either…
await new Promise((resolve) => process.stderr.write("\n", () => resolve(null)));
// …so sleep?
await sleep(100); // 50ms not enough.
```

Maybe I misunderstand what "flush" means inside of Pino here.

Logger is set up like this, in case anyone was wondering:

```ts
const destination = createDestination({ dest: process.stderr.fd, sync: true });

const transport = {
  target: "pino-pretty",
  options: {
    destination: process.stderr.fd,
    colorize: true,
    singleLine: false,
    sync: true,
  },
};
```

I thought maybe `sync: true` would help, but it does not appear like it. Just thinking about:

> can't think of a scenario where you'd _need_ pretty print and async together

Is there some way to not use async here?
